File: solutions/_106_construct_binary_tree_from_inorder_and_postorder_traversal/Solution.java

package _106_construct_binary_tree_from_inorder_and_postorder_traversal;

import java.util.HashMap;

class Solution {
    HashMap<Integer, Integer> map = new HashMap<>();
    int[] postorder;

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        this.postorder = postorder;
        return build(0, inorder.length - 1, 0, postorder.length - 1);
    }

    private TreeNode build(int inStart, int inEnd, int postStart, int postEnd) {
        if (inStart > inEnd || postStart > postEnd) {
            return null;
        }

        TreeNode node = new TreeNode(postorder[postEnd]);
        int inIdx = map.get(node.val);
        int leftSize = inIdx - inStart;

        node.left = build(inStart, inIdx - 1, postStart, postStart + leftSize - 1);
        node.right = build(inIdx + 1, inEnd, postStart + leftSize, postEnd - 1);

        return node;
    }
}

/*
 * https://leetcode.com/problems/construct-binary-tree-from-inorder-and-
 * postorder-traversal/
 */

File: solutions/_106_construct_binary_tree_from_inorder_and_postorder_traversal/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
var buildTree = function (inorder, postorder) {
  const map = new Map();
  for (let i = 0; i < inorder.length; i++) {
    map.set(inorder[i], i);
  }

  const build = (inStart, inEnd, postStart, postEnd) => {
    if (inStart > inEnd || postStart > postEnd) {
      return null;
    }

    const node = new TreeNode(postorder[postEnd]);
    const inIdx = map.get(node.val);
    const leftSize = inIdx - inStart;

    node.left = build(inStart, inIdx - 1, postStart, postStart + leftSize - 1);
    node.right = build(inIdx + 1, inEnd, postStart + leftSize, postEnd - 1);
    return node;
  };

  return build(0, inorder.length - 1, 0, postorder.length - 1);
};

module.exports = { buildTree, TreeNode };

/*
https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/
*/
